Subject: [PATCH]
---
diff --git a/plugin/executable/cache/cache.go b/plugin/executable/cache/cache.go
--- a/plugin/executable/cache/cache.go
+++ b/plugin/executable/cache/cache.go
@@ -106,6 +106,7 @@
 	prefetchFailTotal prometheus.Counter
 	capturedNext      atomic.Pointer[sequence.ChainWalker]
 	captureOnce       sync.Once
+	keyLocks          sync.Map
 }
 
 func Init(bp *coremain.BP, args any) (any, error) {
@@ -231,33 +232,62 @@
 		c.doLazyUpdate(msgKey, qCtx, next)
 	}
 	if cachedResp != nil { // cache hit
-		c.hitTotal.Inc()
-		cachedResp.Id = q.Id // change msg id
-		qCtx.SetResponse(cachedResp)
-		qCtx.FromHosts = false
-		qCtx.FromArbitrary = false
-		if v, _, ok := c.backend.Get(key(msgKey)); ok && v != nil {
-			v.hitCount.Add(1)
-			ttl := int(v.expirationTime.Sub(v.storedTime).Seconds())
-			remainingTtl := int(v.expirationTime.Sub(time.Now()).Seconds())
-			if remainingTtl < 0 {
-				remainingTtl = 0
-			}
-			qCtx.SetCacheState(true, lazyHit, ttl, remainingTtl)
-		} else {
-			qCtx.SetCacheState(true, lazyHit, 0, 0)
-		}
-	} else {
-		qCtx.SetCacheState(false, false, 0, 0)
+		c.onCacheHit(cachedResp, q, qCtx, msgKey, lazyHit)
+		return next.ExecNext(ctx, qCtx)
+	}
+	return c.execNextWithKeyLimit(ctx, qCtx, next, msgKey, q)
+}
+
+func (c *Cache) execNextWithKeyLimit(ctx context.Context, qCtx *query_context.Context, next sequence.ChainWalker, msgKey string, q *dns.Msg) error {
+	var ch chan struct{}
+	if actual, ok := c.keyLocks.Load(msgKey); ok {
+		ch = actual.(chan struct{})
+	} else {
+		actual, _ := c.keyLocks.LoadOrStore(msgKey, make(chan struct{}, 1))
+		ch = actual.(chan struct{})
+	}
+	select {
+	case ch <- struct{}{}:
+		defer func() {
+			<-ch
+			if len(ch) == 0 {
+				c.keyLocks.Delete(msgKey)
+			}
+		}()
+	case <-ctx.Done():
+		return context.Cause(ctx)
+	}
+	doubleCachedResp, _ := getRespFromCache(msgKey, c.backend, c.args.LazyCacheTTL > 0, expiredMsgTtl)
+	if doubleCachedResp != nil {
+		c.onCacheHit(doubleCachedResp, q, qCtx, msgKey, false)
+		return next.ExecNext(ctx, qCtx)
+	}
+	qCtx.SetCacheState(false, false, 0, 0)
+	err := next.ExecNext(ctx, qCtx)
+	if r := qCtx.R(); r != nil {
+		saveRespToCache(msgKey, r, c.backend, c.args.LazyCacheTTL)
+		c.updatedKey.Add(1)
+	}
+	return err
+}
+
+func (c *Cache) onCacheHit(cachedResp *dns.Msg, q *dns.Msg, qCtx *query_context.Context, msgKey string, lazyHit bool) {
+	c.hitTotal.Inc()
+	cachedResp.Id = q.Id // change msg id
+	qCtx.SetResponse(cachedResp)
+	qCtx.FromHosts = false
+	qCtx.FromArbitrary = false
+	if v, _, ok := c.backend.Get(key(msgKey)); ok && v != nil {
+		v.hitCount.Add(1)
+		ttl := int(v.expirationTime.Sub(v.storedTime).Seconds())
+		remainingTtl := int(v.expirationTime.Sub(time.Now()).Seconds())
+		if remainingTtl < 0 {
+			remainingTtl = 0
+		}
+		qCtx.SetCacheState(true, lazyHit, ttl, remainingTtl)
+	} else {
+		qCtx.SetCacheState(true, lazyHit, 0, 0)
 	}
-
-	err := next.ExecNext(ctx, qCtx)
-
-	if r := qCtx.R(); r != nil && cachedResp != r { // pointer compare. r is not cachedResp
-		saveRespToCache(msgKey, r, c.backend, c.args.LazyCacheTTL)
-		c.updatedKey.Add(1)
-	}
-	return err
 }
 
 // doLazyUpdate starts a new goroutine to execute next node and update the cache in the background.
